File: training/src/latentlab/planning/cem.py
"""Cross-Entropy Method planner over action sequences in latent space.

Plan: sample a population of action sequences from a per-step Gaussian,
roll each through the PREDICTOR ONLY (no env, no decoder — pure latent
imagination), score by distance between the imagined final latent and the
encoded goal latent, refit the Gaussian on the elite set, repeat.

Wall avoidance falls out of the world model: the predictor learned that
actions pushing into the wall don't move the embedding, so sequences that
try to tunnel score badly and CEM discovers the doorway route — nobody
hard-coded the wall into the planner.

MPC usage: call ``plan`` each env step, execute the first action, and the
planner warm-starts the next call by shifting its mean one step forward.
This class is the reference implementation for the TypeScript port
(Phase 6); keep the algorithm boring and the batching predictor-friendly.
"""

from __future__ import annotations

from dataclasses import dataclass

import torch

from latentlab.models.predictor import Predictor


@dataclass(frozen=True)
class CEMConfig:
    """Planner hyperparameters (defaults tuned for Two Rooms)."""

    horizon: int = 12  # imagined steps; ~1 room-width at max_step
    population: int = 256  # candidates per iteration
    iterations: int = 4  # CEM refinement rounds
    elite_frac: float = 0.125  # top 32 of 256
    init_sigma: float = 0.05  # exploration noise ~ 0.6 * max_step
    min_sigma: float = 0.01  # sigma floor: never stop exploring entirely
    action_bound: float = 0.08  # env max_step; actions clamped to +/- this
    # Cost = sum of latent distance over ALL imagined steps, not just the
    # endpoint. Endpoint-only cost makes the first action underdetermined
    # whenever the goal is reachable in < horizon steps (any wander-then-
    # arrive sequence ties), which turns MPC into a random walk. Dense cost
    # rewards immediate progress and pins the first action down.
    terminal_weight: float = 4.0  # extra weight on the final step: arrive AND stay


class CEMPlanner:
    """Stateful (warm-started) CEM planner in latent space."""

    def __init__(
        self,
        predictor: Predictor,
        config: CEMConfig | None = None,
        generator: torch.Generator | None = None,
    ) -> None:
        self.predictor = predictor
        self.config = config if config is not None else CEMConfig()
        self.device = next(predictor.parameters()).device
        self.generator = generator
        self._mean: torch.Tensor | None = None  # (H, 2) warm-start between MPC steps

    def reset(self) -> None:
        """Clear the warm-start (call at the start of each episode)."""
        self._mean = None

    @torch.no_grad()
    def rollout_costs(
        self, z0: torch.Tensor, actions: torch.Tensor, z_goal: torch.Tensor
    ) -> torch.Tensor:
        """Dense trajectory cost of (P, H, 2) action sequences from latent z0 (D,).

        cost_p = sum_t ||z_t - z_goal||^2 + (terminal_weight - 1) * ||z_H - z_goal||^2
        """
        z = z0.unsqueeze(0).expand(actions.shape[0], -1).contiguous()
        horizon = actions.shape[1]
        costs = torch.zeros(actions.shape[0], device=z0.device)
        for t in range(horizon):
            z = self.predictor(z, actions[:, t])
            step_cost = ((z - z_goal) ** 2).sum(dim=1)
            weight = self.config.terminal_weight if t == horizon - 1 else 1.0
            costs += weight * step_cost
        return costs

    @torch.no_grad()
    def plan(self, z0: torch.Tensor, z_goal: torch.Tensor) -> tuple[torch.Tensor, dict[str, float]]:
        """One MPC planning call.

        Args:
            z0: (D,) current latent. z_goal: (D,) goal latent.

        Returns:
            actions: (H, 2) best action sequence found (execute actions[0]).
            info: best cost and final sigma (for latency/telemetry panels).
        """
        cfg = self.config
        n_elite = max(1, int(cfg.population * cfg.elite_frac))

        if self._mean is None:
            mean = torch.zeros(cfg.horizon, 2, device=self.device)
        else:
            # Warm start: shift last plan one step forward, repeat final row.
            mean = torch.cat([self._mean[1:], self._mean[-1:]], dim=0)
        sigma = torch.full((cfg.horizon, 2), cfg.init_sigma, device=self.device)

        best_actions = mean.clone()
        best_cost = float("inf")
        for _ in range(cfg.iterations):
            noise = torch.randn(
                cfg.population, cfg.horizon, 2, device=self.device, generator=self.generator
            )
            candidates = (mean + noise * sigma).clamp(-cfg.action_bound, cfg.action_bound)
            # Keep the current mean in the population so refinement is monotone.
            candidates[0] = mean.clamp(-cfg.action_bound, cfg.action_bound)

            costs = self.rollout_costs(z0, candidates, z_goal)
            elite_idx = costs.topk(n_elite, largest=False).indices
            elites = candidates[elite_idx]

            mean = elites.mean(dim=0)
            sigma = elites.std(dim=0).clamp_min(cfg.min_sigma)

            iter_best = int(elite_idx[0].item())
            iter_best_cost = float(costs[iter_best].item())
            if iter_best_cost < best_cost:
                best_cost = iter_best_cost
                best_actions = candidates[iter_best].clone()

        self._mean = best_actions
        return best_actions, {"best_cost": best_cost, "mean_sigma": float(sigma.mean().item())}


if __name__ == "__main__":
    # Standalone sanity: toy 2-D "latent" where z' = z + a (identity world
    # model). CEM must steer toward the goal.
    torch.manual_seed(0)

    class IdentityPredictor(Predictor):
        def __init__(self) -> None:
            super().__init__(latent_dim=2)

        def forward(self, z: torch.Tensor, action: torch.Tensor) -> torch.Tensor:
            return z + action

    planner = CEMPlanner(IdentityPredictor(), CEMConfig(horizon=8))
    z0 = torch.tensor([0.0, 0.0])
    z_goal = torch.tensor([0.4, -0.2])
    actions, info = planner.plan(z0, z_goal)
    endpoint = z0 + actions.sum(dim=0)
    print(f"goal {z_goal.tolist()} -> planned endpoint {[round(v, 3) for v in endpoint.tolist()]}")
    error = (endpoint - z_goal).norm().item()
    print(f"endpoint error {error:.4f} (dense cost also charges the approach path,")
    print("so best_cost is not ~0 even for a perfect plan; endpoint error should be small)")
    assert error < 0.05, "CEM failed to steer the identity world model to the goal"

File: web/src/planner/cem.ts
/**
 * CEM planner in latent space — TypeScript port of
 * training/src/latentlab/planning/cem.py (the reference implementation).
 *
 * Key property carried over: the cost is DENSE — the sum of latent distance
 * to goal over ALL imagined steps plus an extra-weighted terminal step.
 * Endpoint-only cost leaves the first action underdetermined whenever the
 * goal is reachable in fewer than `horizon` steps, which turns MPC into a
 * random walk (this failed at 20% success in Python before the fix; see the
 * Python module docstring).
 *
 * The predictor is abstracted as an async batched function so the planner is
 * unit-testable without ONNX and runs identically in the worker.
 */

export interface CEMConfig {
  horizon: number
  population: number
  iterations: number
  eliteFrac: number
  initSigma: number
  minSigma: number
  actionBound: number
  terminalWeight: number
}

export const DEFAULT_CEM_CONFIG: CEMConfig = {
  horizon: 12,
  population: 256,
  iterations: 4,
  eliteFrac: 0.125,
  initSigma: 0.05,
  minSigma: 0.01,
  actionBound: 0.08,
  terminalWeight: 4.0,
}

/** Batched one-step world model: (latents B*D, actions B*2, B) -> latents B*D. */
export type PredictFn = (
  latents: Float32Array,
  actions: Float32Array,
  batch: number,
) => Promise<Float32Array>

export interface PlanResult {
  /** Best action sequence (horizon * 2), execute [0], [1] first. */
  actions: Float32Array
  bestCost: number
  /** Imagined latent trajectory of the best plan (horizon * latentDim). */
  bestTrajectory: Float32Array
  /** Elite imagined trajectories for viz (nElites * horizon * latentDim). */
  eliteTrajectories: Float32Array
  nElitesReturned: number
}

/** mulberry32 PRNG — deterministic across platforms, good enough for CEM. */
export function mulberry32(seed: number): () => number {
  let a = seed >>> 0
  return () => {
    a = (a + 0x6d2b79f5) | 0
    let t = Math.imul(a ^ (a >>> 15), 1 | a)
    t = (t + Math.imul(t ^ (t >>> 7), 61 | t)) ^ t
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296
  }
}

/** Box-Muller standard normal from a uniform source. */
function gaussianPair(rand: () => number): [number, number] {
  let u = 0
  while (u === 0) u = rand()
  const v = rand()
  const r = Math.sqrt(-2 * Math.log(u))
  const theta = 2 * Math.PI * v
  return [r * Math.cos(theta), r * Math.sin(theta)]
}

export class CEMPlanner {
  private mean: Float32Array | null = null // (H*2) warm start
  private rand: () => number

  constructor(
    private predict: PredictFn,
    private latentDim: number,
    public config: CEMConfig = { ...DEFAULT_CEM_CONFIG },
    seed = 0,
  ) {
    this.rand = mulberry32(seed)
  }

  reset(): void {
    this.mean = null
  }

  /** One MPC planning call. z0/zGoal are (latentDim,) latents. */
  async plan(z0: Float32Array, zGoal: Float32Array, vizElites = 8): Promise<PlanResult> {
    const { horizon: H, population: P, iterations, eliteFrac } = this.config
    const { initSigma, minSigma, actionBound, terminalWeight } = this.config
    const D = this.latentDim
    const nElite = Math.max(1, Math.floor(P * eliteFrac))

    // Warm start: shift the previous plan one step, repeat the last row.
    let mean = new Float32Array(H * 2)
    if (this.mean) {
      mean.set(this.mean.subarray(2), 0)
      mean[H * 2 - 2] = this.mean[H * 2 - 2]
      mean[H * 2 - 1] = this.mean[H * 2 - 1]
    }
    const sigma = new Float32Array(H * 2).fill(initSigma)

    const candidates = new Float32Array(P * H * 2)
    const costs = new Float64Array(P)
    // Trajectory latents for the current iteration: (P, H, D) — reused.
    const traj = new Float32Array(P * H * D)
    const stepLatents = new Float32Array(P * D)
    const stepActions = new Float32Array(P * 2)

    let bestCost = Infinity
    const bestActions = new Float32Array(H * 2)
    const bestTrajectory = new Float32Array(H * D)
    let eliteIdx: number[] = []

    for (let iter = 0; iter < iterations; iter++) {
      // Sample candidates ~ N(mean, sigma), clamped. Candidate 0 = mean.
      for (let p = 0; p < P; p++) {
        for (let k = 0; k < H * 2; k += 2) {
          if (p === 0) {
            candidates[p * H * 2 + k] = clamp(mean[k], actionBound)
            candidates[p * H * 2 + k + 1] = clamp(mean[k + 1], actionBound)
          } else {
            const [g0, g1] = gaussianPair(this.rand)
            candidates[p * H * 2 + k] = clamp(mean[k] + g0 * sigma[k], actionBound)
            candidates[p * H * 2 + k + 1] = clamp(mean[k + 1] + g1 * sigma[k + 1], actionBound)
          }
        }
      }

      // Batched rollout with dense cost accumulation.
      costs.fill(0)
      for (let p = 0; p < P; p++) stepLatents.set(z0, p * D)
      for (let t = 0; t < H; t++) {
        for (let p = 0; p < P; p++) {
          stepActions[p * 2] = candidates[p * H * 2 + t * 2]
          stepActions[p * 2 + 1] = candidates[p * H * 2 + t * 2 + 1]
        }
        const next = await this.predict(stepLatents, stepActions, P)
        stepLatents.set(next)
        const weight = t === H - 1 ? terminalWeight : 1.0
        for (let p = 0; p < P; p++) {
          let dist = 0
          for (let d = 0; d < D; d++) {
            const diff = next[p * D + d] - zGoal[d]
            dist += diff * diff
          }
          costs[p] += weight * dist
          traj.set(next.subarray(p * D, (p + 1) * D), p * H * D + t * D)
        }
      }

      // Elites: indices of the nElite lowest costs.
      eliteIdx = argsortAscending(costs).slice(0, nElite)

      // Refit mean/sigma from elites.
      mean = new Float32Array(H * 2)
      for (const p of eliteIdx) {
        for (let k = 0; k < H * 2; k++) mean[k] += candidates[p * H * 2 + k]
      }
      for (let k = 0; k < H * 2; k++) mean[k] /= nElite
      for (let k = 0; k < H * 2; k++) {
        let variance = 0
        for (const p of eliteIdx) {
          const diff = candidates[p * H * 2 + k] - mean[k]
          variance += diff * diff
        }
        // Match torch.std: unbiased (n-1) sample std.
        sigma[k] = Math.max(Math.sqrt(variance / Math.max(nElite - 1, 1)), minSigma)
      }

      const iterBest = eliteIdx[0]
      if (costs[iterBest] < bestCost) {
        bestCost = costs[iterBest]
        bestActions.set(candidates.subarray(iterBest * H * 2, (iterBest + 1) * H * 2))
        bestTrajectory.set(traj.subarray(iterBest * H * D, (iterBest + 1) * H * D))
      }
    }

    this.mean = bestActions.slice()

    // Elite trajectories from the FINAL iteration for candidate viz.
    const nReturn = Math.min(vizElites, eliteIdx.length)
    const eliteTrajectories = new Float32Array(nReturn * H * D)
    for (let i = 0; i < nReturn; i++) {
      const p = eliteIdx[i]
      eliteTrajectories.set(traj.subarray(p * H * D, (p + 1) * H * D), i * H * D)
    }

    return {
      actions: bestActions,
      bestCost,
      bestTrajectory,
      eliteTrajectories,
      nElitesReturned: nReturn,
    }
  }
}

function clamp(v: number, bound: number): number {
  return Math.min(Math.max(v, -bound), bound)
}

function argsortAscending(values: Float64Array): number[] {
  return [...values.keys()].sort((a, b) => values[a] - values[b])
}
